gapis/api/vulkan/api/coherent_memory.api

```diff
@@ -56,58 +56,170 @@ sub void readCoherentMemory(ref!DeviceMemoryObject memory, VkDeviceSize readOffs
   }
 }
 
-sub void readCoherentMemoryInBuffer(ref!BufferObject buffer, VkDeviceSize readOffset, VkDeviceSize readSize) {
+sub void readMemoryInBuffer(ref!BufferObject buffer, VkDeviceSize readOffset, VkDeviceSize readSize) {
   mem := buffer.Memory
   if mem != null {
-    if as!u64(readSize) == 0xFFFFFFFFFFFFFFFF {
-      readCoherentMemory(mem, buffer.MemoryOffset + readOffset, buffer.Info.Size - readOffset)
-    } else {
-      readCoherentMemory(mem, buffer.MemoryOffset + readOffset, readSize)
+    readBegin := buffer.MemoryOffset + readOffset
+    actualReadSize := switch as!u64(readSize) {
+      case 0xFFFFFFFFFFFFFFFF:
+        buffer.Info.Size - readOffset
+      default:
+        readSize
+    }
+    readEnd := readBegin + actualReadSize
+    readCoherentMemory(mem, readBegin, actualReadSize)
+    if len(mem.Data) != 0 {
+      read(mem.Data[readBegin : readEnd])
     }
   }
 }
 
-sub void readCoherentMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings) {
-  for _, _, v in bufferBindings {
+sub void readMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings, map!(u32, VkDeviceSize) bufferBindingOffsets) {
+  for k, _, v in bufferBindings {
```

Comment:
Not sure if the logic is wrong here, but the triple is index, key, value. `k` is an odd variable choice for index, unless `k` always equals `i`.

Reply:
I think @ben-clayton caught a very important point! It looks like what we want here is `descriptor array index`, so we should use `for _, k, v in bufferBindings {`.

I think there may be similar spots around? And I think we should try it through all the samples and tests we have?

Reply:
The `k` name is due to an earlier version in which the loops in `readWriteMemoryInBoundDescriptorSets` (`i`), `readMemoryInDescriptorSet` (`j`), and `readMemoryInBufferBindings` (`k`) were nested range loops, similar to `dovkCmdBindDescriptorSets`.

I think you're right that using the map iteration index is wrong here. Since this uses Go maps under the hood, the iteration order is undefined, so I think the the index `k` is uncorrelated with the value `v`, correct?

I will switch these cases to `for i in (0 .. len(bufferBindings))` to match the pattern in `dovkCmdBindDescriptorSets` and emphasize that the keys are contiguous.

Reply:
I found what appear to be similar issues in `dovkCmdBindDescriptorSets`
```
  for i , _ , set in args.DescriptorSets {
    lastDrawInfo().DescriptorSets[args.FirstSet + as!u32(i)] = DescriptorSets[set]
  }
```
and `dovkCmdBindVertexBuffers`
```
  for i , _ , v in bind.Buffers ...
```
In each of these cases, it looks like the index `i` and the value, `set` or `v`, each range over the correct set, but in uncorrelated order, yes? If so I will include a fix of these together with the instances I introduced.